### ISSUE TITLE
cannot install packages anymore

### ISSUE THREAD

**Issue by ilanschnell** [conda/conda#2061]
I'm on master:
```
$ conda --version
conda 3.19.1-63-gaaafb26
```
I believe this has something to do with the refactored solved @mcg1969 wrote.  Using `3.19.1`, I had no problems, but when I switch to `master`, I get:
```
$ conda install -c ilan foo
Fetching package metadata: ......
Solving package specifications: .
Error: Unsatisfiable package specifications.
Generating hint: 
[      COMPLETE      ]|###################################################| 100%


The following specifications were found to be in conflict:
  - spacy (target=spacy-0.68-np19py27_0.tar.bz2)
  - foo
  - gevent_zeromq (target=gevent_zeromq-0.2.5-py27_3.tar.bz2)
  - into (target=into-0.1.3-np19py27_0.tar.bz2)
  - mayavi (target=mayavi-4.4.0-np19py27_0.tar.bz2)
  - opencv (target=opencv-3.0.0-np110py27_0.tar.bz2)
  - bottleneck (target=bottleneck-1.0.0-np110py27_0.tar.bz2)
  - holoviews (target=holoviews-1.3.0-np19py27_0.tar.bz2)
Use "conda info <package>" to see the dependencies for each package.
```
The package I'm trying to install (`foo`) has no dependencies other than `python`.  Why is solver looking at all these other packages I have installed, which are unrelated?

I realize that my state of packages might be in some inconsistent state (and I could clean this up).  However, we cannot expect all conda users to bring their environments into a consistent state, before they can install anything.  IMHO, this is a showstopper for releasing conda.

**Comment by groutr** [comment]
@mcg1969

**Comment by mcg1969** [comment]
Got it. I agree this is a showstopper. We need to figure out a solution here. But as I told you in person, it is _conda's fault_ that this happened!

**Comment by ilanschnell** [comment]
I'm suggesting the following:
If the conda environment is in a consistent state, conda should keep it consistent and do the hole thing.
But if the conda is in an inconsistent state, conda should just take the package to be installed into consideration (basically what it had been doing all along).

**Comment by asmeurer** [comment]
Taking into account packages that are already installed is a bad plan unless the solver has some way to indicate packages to be removed (for example, "remove `anaconda` so that I can update `python`").

**Comment by mcg1969** [comment]
Taking into account packages that are already installed is the only way to ensure a consistent state after a `conda install`, `conda remove`, or `conda update`. Moving forward we're not allowing `conda` to break a dependency-compliant environment with these commands unless you do a `--force`.

The problem here, of course, is that we _have_ been allowing `conda` to break environments, silently, without warning, so people will be surprised when suddenly `conda` tries to be strict. So we're going to have to ease people into this new regime by being intelligent about these cases.

New environments, of course, will start out consistent and stay that way. So this is a legacy problem.

**Comment by asmeurer** [comment]
@mcg1969 I think that's great. I've been wanting to do this for a long time (https://github.com/conda/conda/pull/634 https://github.com/conda/conda/issues/454), and have mostly been blocked on a slow solver. **But**, as I noted, you can't impose this restriction unless the solver is able to solve for removed packages. As a trivial example, as soon as someone `conda remove`s a package that is a dependency their environment will become inconsistent (albeit typically in a trivial way). 

As a more complex, but also quite common example, someone who installs anaconda and wants to update, say, Python to the next minor version. Right now anaconda 2.5.0 depends on Python 3.5.1. When Python 3.5.2 comes out if someone who installed anaconda runs `conda update python`, under this scheme, it won't update it. If they run `conda install python=3.5.2` it will give unsatisfiable package specifications of `python` and `anaconda`. The only way for this to be completely smooth, from a user point of view, is if the solver can realize that `anaconda` is not important and can be removed. `anaconda` is a simple, and common example, but there are others as well. Typical examples are now outdated packages that were installed automatically at some point, but are not needed any more, but would restrict what can be installed due to their dependencies (say, a package that is no longer supported and was only built against an older version of NumPy).

Finally, some people "work around" inconsistent dependencies by doing the install in two stages (e.g., a package isn't built against the version of numpy you need, but the build that exists turns out to work just fine, so you install and then manually change the numpy version from under it).

If this is already done, let me know (I haven't been following the new solver developments very closely). 

My suggestions if it isn't:

- Guess what packages are important to a user from the history file.
- Allow a "version" of a package to be installed in the solver to be "removed" (or "not installed"). It's an exercise for the reader as to what priority that should have over the other versions. 
- Only enable this feature optionally, at least at first (perhaps on a per-environment basis). But don't enable it by default, until you have a grounding on what the implications are.

**Comment by mcg1969** [comment]
Oh, I'm totally overhauling the solver, @asmeurer. We will indeed be handling `conda remove` properly. There will be a dev build with this functionality next week.

**Comment by mcg1969** [comment]
@ilanschnell, if you get a chance to check the `mgrant-remove` branch let me know.

**Comment by kalefranz** [comment]
Sounds like this was resolved with `4.0`.
